### PR TITLE
Fix BuildConfigInstantiateFailed warning when lastVersion == 0

### LINKED ISSUE #16557: Always see BuildConfigInstantiateFailed warning event after creating build config
I always see this warning after creating a build config using `oc new-app` or through the web console:

> BuildConfigInstantiateFailed
> gave up on Build for BuildConfig test/nodejs-ex (0) due to fatal error: the LastVersion(1) on build config test/nodejs-ex does not match the build request LastVersion(0)

This doesn't seem to cause a problem except that it adds noise to the events and might make the user think something's wrong. Note that this event will show up in the notifications drawer in the web console, so it's somewhat more prominent there than on the CLI.

##### Version

oc v3.7.0-alpha.1+75c64e3-653
kubernetes v1.7.0+80709908fd
features: Basic-Auth

Server https://127.0.0.1:8443
openshift v3.7.0-alpha.1+899d4c9-676
kubernetes v1.7.0+80709908fd

##### Steps To Reproduce

```
$ oc new-app https://github.com/openshift/ruby-ex.git
$ oc describe bc/ruby-ex
```

You see the `BuildConfigInstantiateFailed` warning for the build config.

### PR DESCRIPTION
**Pull request by coreydaley**
Build config instatiate should not emit a warning when
bc.Status.LastVersion != lastVersion and lastVersion == 0
Fixes #16557

### CODE REVIEW

**Comment by coreydaley** [comment]
@openshift/sig-developer-experience ptal

**Comment by bparees** [comment]
I don't think this is the right fix.  You've not just eliminated the warning, isn't this now going to run two builds (the configchangetriggered one and the imagechangetriggered one)?

**Comment by bparees** [comment]
@coreydaley did you forget about this?

**Comment by coreydaley** [comment]
@bparees nope, it's next on my list

**Comment by coreydaley** [comment]
Not sure why this got closed?

**Comment by bparees** [comment]
Did you delete the branch? That will do it.

Ben Parees | OpenShift

**Comment by coreydaley** [comment]
I reset my branch so that I could start over on this PR and pushed it, weird that it closed, reopening with new code.

**Comment by coreydaley** [comment]
@openshift/sig-developer-experience ptal

File: pkg/build/controller/buildconfig/buildconfig_controller.go [hunk 1]
```diff
@@ -136,7 +136,12 @@ func (c *BuildConfigController) handleBuildConfig(bc *buildapi.BuildConfig) erro
 		} else if buildgenerator.IsFatal(err) || kerrors.IsNotFound(err) || kerrors.IsBadRequest(err) || kerrors.IsForbidden(err) {
 			instantiateErr = fmt.Errorf("gave up on Build for BuildConfig %s due to fatal error: %v", bcDesc(bc), err)
 			utilruntime.HandleError(instantiateErr)
-			c.recorder.Event(bc, kapi.EventTypeWarning, "BuildConfigInstantiateFailed", instantiateErr.Error())
+			// Don't emit this event if the BuildRequest.LastVersion equals zero
+			// Caused by a race condition between the ImageChangeTrigger vs the BuildConfigChangeTrigger
+			// Fixes https://github.com/openshift/origin/issues/16557
+			if *request.LastVersion != int64(0) {
```
Comment (bparees) [inline pkg/build/controller/buildconfig/buildconfig_controller.go]: request.LastVersion will always be zero in this codepath.  so you're going to mask any/all errors.

I think you need to actually look at the error you got from Instantiate (and/or make Instantiate throw a more distinctive error you can look for)

**Comment by coreydaley** [comment]
@bparees ptal

File: pkg/build/controller/buildconfig/buildconfig_controller.go [hunk 2]
```diff
@@ -136,7 +137,11 @@ func (c *BuildConfigController) handleBuildConfig(bc *buildapi.BuildConfig) erro
 		} else if buildgenerator.IsFatal(err) || kerrors.IsNotFound(err) || kerrors.IsBadRequest(err) || kerrors.IsForbidden(err) {
 			instantiateErr = fmt.Errorf("gave up on Build for BuildConfig %s due to fatal error: %v", bcDesc(bc), err)
 			utilruntime.HandleError(instantiateErr)
-			c.recorder.Event(bc, kapi.EventTypeWarning, "BuildConfigInstantiateFailed", instantiateErr.Error())
+			// Fixes https://github.com/openshift/origin/issues/16557
+			// Caused by a race condition between the ImageChangeTrigger vs the BuildConfigChangeTrigger
+			if !strings.Contains(instantiateErr.Error(), fmt.Sprintf("the LastVersion(1) on build config %s/%s does not match the build request LastVersion(0)", bc.Namespace, bc.Name)) {
+				c.recorder.Event(bc, kapi.EventTypeWarning, "BuildConfigInstantiateFailed", instantiateErr.Error())
+			}
```
Comment (gabemontero) [inline pkg/build/controller/buildconfig/buildconfig_controller.go]: What if a single request comes through (i.e. not a race between ICT and CCT) after the initial BC create with a bad last version?  Wouldn't we be masking a legitimate error in that case?

Unfortunately, don't know if I have a super great solution ... perhaps we fetch the BC before performing this filter and see if it was *recently created* ... where determining what is *recent* is perhaps kludgy ... but maybe at least an improvement.

Dumb non-transactional databases

**Comment by coreydaley** [comment]
@jim-minter any thoughts?

File: pkg/build/controller/buildconfig/buildconfig_controller.go (hunk 2 above)
Comment (bparees) [reply pkg/build/controller/buildconfig/buildconfig_controller.go]: > What if a single request comes through (i.e. not a race between ICT and CCT) after the initial BC create with a bad last version? 

request comes from where? to where?  this logic is all in the buildconfig controller.  There are no other buildrequests its going to process.

@coreydaley your check on LastVersion(1) is too specific.  I could potentially start multiple builds before the buildconfig controller puts in its request.  Then the error would be "the LastVersion(5) on build config..." or whatever.

Just check for the back half: `does not match the build request LastVersion(0)`

**Comment by gabemontero** [comment]
On Mon, Feb 19, 2018 at 4:59 PM, Ben Parees <notifications@github.com>
wrote:

> *@bparees* commented on this pull request.
> ------------------------------
>
> In pkg/build/controller/buildconfig/buildconfig_controller.go
> <https://github.com/openshift/origin/pull/17146#discussion_r169179068>:
>
> > @@ -136,7 +137,11 @@ func (c *BuildConfigController) handleBuildConfig(bc *buildapi.BuildConfig) erro
>  		} else if buildgenerator.IsFatal(err) || kerrors.IsNotFound(err) || kerrors.IsBadRequest(err) || kerrors.IsForbidden(err) {
>  			instantiateErr = fmt.Errorf("gave up on Build for BuildConfig %s due to fatal error: %v", bcDesc(bc), err)
>  			utilruntime.HandleError(instantiateErr)
> -			c.recorder.Event(bc, kapi.EventTypeWarning, "BuildConfigInstantiateFailed", instantiateErr.Error())
> +			// Fixes https://github.com/openshift/origin/issues/16557
> +			// Caused by a race condition between the ImageChangeTrigger vs the BuildConfigChangeTrigger
> +			if !strings.Contains(instantiateErr.Error(), fmt.Sprintf("the LastVersion(1) on build config %s/%s does not match the build request LastVersion(0)", bc.Namespace, bc.Name)) {
> +				c.recorder.Event(bc, kapi.EventTypeWarning, "BuildConfigInstantiateFailed", instantiateErr.Error())
> +			}
>
> What if a single request comes through (i.e. not a race between ICT and
> CCT) after the initial BC create with a bad last version?
>
> request comes from where? to where? this logic is all in the buildconfig
> controller. There are no other buildrequests its going to process.
>
So you are saying somebody couldn't manually hit the build config rest
endpoint with a HTTP req and interfere with this?

**Comment by bparees** [comment]
> So you are saying somebody couldn't manually hit the build config rest endpoint with a HTTP req and interfere with this?

interfere with this how?

**Comment by coreydaley** [comment]
@bparees update pushed

**Comment by gabemontero** [comment]
On Mon, Feb 19, 2018 at 5:14 PM Ben Parees <notifications@github.com> wrote:

> So you are saying somebody couldn't manually hit the build config rest
> endpoint with a HTTP req and interfere with this?
>
> interfere with this how?
>
create a request that is independent of the concurrent requests @coreydaley
noted coming from the triggers

And when the say the human manually builds ths json for the request he uses
the wrong last version

**Comment by bparees** [comment]
> And when the say the human manually builds ths json for the request he uses
the wrong last version

the code that handles that buildrequest is not the codepath that @coreydaley added his error checking/filtering to.  (both codepaths call instantiate, but only the buildconfigcontroller codepath is having the filtering added to it)

**Pull request merged**